File: SICP/Lecture_1/Code/Java/Peano/src/fayelab/sicp/peano/Peano.java
package fayelab.sicp.peano;

public class Peano
{
    public static int iterationAdd(int x, int y)
    {
        if(x == 0)
        {
            return y;
        }
        
        return iterationAdd(minusOne(x), plusOne(y));
    }
    
    public static int recursiveAdd(int x, int y)
    {
        if(x == 0)
        {
            return y;
        }
        
        return plusOne(recursiveAdd(minusOne(x), y));
    }
    
    private static int plusOne(int x)
    {
        return x + 1;
    }
    
    private static int minusOne(int x)
    {
        return x - 1;
    }
}

File: SICP/Lecture_1/Code/Python/Peano/peano.py
def iteration_add(x, y):
    if x == 0:
        return y

    return iteration_add(minus_one(x), plus_one(y))

def recursive_add(x, y):
    if x == 0:
        return y

    return plus_one(recursive_add(minus_one(x), y))

def minus_one(x):
    return x - 1

def plus_one(x):
    return x + 1